src/java_bytecode/library/src/java/lang/Class.java

```diff
@@ -0,0 +1,423 @@
+     * not null and the caller's class loader is not the same as or an ancestor of
+     * the class loader for the class whose class loader is requested, then
+     * this method calls the security manager's {@code checkPermission}
+     * method with a {@code RuntimePermission("getClassLoader")}
+     * permission to ensure it's ok to access the class loader for the class.
+     *
+     * <p>If this object
+     * represents a primitive type or void, null is returned.
+     *
+     * @return  the class loader that loaded the class or interface
+     *          represented by this object.
+     * @throws SecurityException
+     *    if a security manager exists and its
+     *    {@code checkPermission} method denies
+     *    access to the class loader for the class.
+     * @see java.lang.ClassLoader
+     * @see SecurityManager#checkPermission
+     * @see java.lang.RuntimePermission
+     */
+    @CallerSensitive
+    public ClassLoader getClassLoader() {
+        ClassLoader cl = getClassLoader0();
+        if (cl == null)
+            return null;
+        SecurityManager sm = System.getSecurityManager();
+        if (sm != null) {
+            ClassLoader.checkClassLoaderPermission(cl, Reflection.getCallerClass());
+        }
+        return cl;
+    }
+
+    ClassLoader getClassLoader0() {
+        // DIFFBLUE MODEL LIBRARY The real java.lang.Class stores a
+        // `private final ClassLoader classLoader` which is initialised by the
+        // jvm rather than by the constructor of this object.
+        // TODO test-gen should understand this method natively.
+        return null;
+    }
+
+    public String getSimpleName() {
+        /* DIFFBLUE MODEL LIBRARY TODO: No implementation for getComponentType
+        if (isArray())
+            return getComponentType().getSimpleName()+"[]";
+        */
+
+        // DIFFBLUE MODEL LIBRARY: instead of calling getSimpleBinaryName()
+        // we inline function calls and simplify
+        String name = getName();
+        int index = name.lastIndexOf('$');
+        if(index == -1) { // top level class
+            return name.substring(name.lastIndexOf('.') + 1); // strip the package name
+        }
+        else {
+            // DIFFBLUE MODEL LIBRARY: in the original JDK getSimpleBinary
+            // looks for "$1", instead we looked for '$' and assume the next
+            // character will be '1'
+            CProver.assume(name.charAt(index+1) == '1');
+            // DIFFBLUE MODEL LIBRARY: $1 should be preceded by a class name
+            CProver.assume(index >= 1);
+            // DIFFBLUE MODEL LIBRARY: in the original JDK getSimpleName
+            // removes the digits that follow
+            CProver.assume(!isAsciiDigit(name.charAt(index + 2)));
+            return name.substring(index + 2);
+        }
+    }
+
+    private static boolean isAsciiDigit(char c) {
+        return '0' <= c && c <= '9';
+    }
+
+    public Class getEnclosingClass() {
```

Comment:
Judging by the weird indentation, this doesn't look like a method from the original JDK. I recommend prefixing such methods with `cprover...` so that we can distinguish them from original methods.
I gather that you imported these classes from models-library in the state they were, but you might want to clean them up a bit, as some of them are not in a particularly good state.

Reply:
On the other hand, we don't want to have too much diff between the core classes and the models-library, so any clean-up should be mirrored in the models-library, but that might be asking a bit much.

Reply:
Same comment, we have just copied the file. Do you want us to proceed with those changes only here?